### ISSUE TITLE
Alternative to typeDfltVal for GOOL's Swift renderer

### ISSUE THREAD

**Issue by bmaclach** [JacquesCarette/Drasil#2200]
I implemented a function in GOOL's swift renderer, `typeDfltVal`, which returns a default value based on a given type. Using default values in this way is something we would like to avoid. The reason for this function is because all Swift state variables require an initial value, but GOOL's `stateVar` function does not accept a value as a parameter because other, non-Swift languages, do not need initial values. We should think about whether we can remove `typeDfltVal` and use some other solution. See further related discussion here: https://github.com/JacquesCarette/Drasil/pull/2197#commitcomment-40209217

**Comment by JacquesCarette** [comment]
The best solution would be for `stateVar` to require a default / starting value.  That would force examples "upstream" to be more specific, which is a good thing.

**Comment by balacij** [comment]
Out of curiosity, why is using default values for the generated Swift code something we want to avoid?

**Comment by JacquesCarette** [comment]
The "bad" idea is to assume that all types have a distinguished 'point' that is safe to use as a default value. If nothing else, this leads to weird, really hard to debug code because the provenance of some values is not clear (i.e. it looks like it appears out of thin air).

Worse: when we allow user-defined types, this would likely break again, as we wouldn't know how to pick a 'default' value.

So the best scheme is to force all variables to have a starting value that has been decided on by 'someone' in a traceable way.

**Comment by B-rando1** [comment]
Does it make sense to have `stateVar` require a default value that's only used by the Swift renderer? Alternatively we could remove it entirely in favour of `stateVarDec`. I can see the argument for either. I guess not defining the variable is idiomatic for most languages, so maybe it is worth keeping.

**Comment by JacquesCarette** [comment]
While not defining variables might be idiomatic for far too many languages, we would rather not do that, in the sense that we internally should always have an 'initial value' for all variables, even if we end up choosing to not render them in some cases.  Note that I did not say 'default value' (that's usually a hack).

So only having `stateVarDec` seems like the principled thing -- but that's not necessarily the only implementation.

**Comment by B-rando1** [comment]
I made a [comment](https://github.com/JacquesCarette/Drasil/pull/5339#issuecomment-5027167109) in #5339. What might be most idiomatic is to just make sure that the programmer has initialized a variable before they use it. This is good practice in any language and will make Swift happy. It'll take a bit of work for GOOL to be able to catch that sort of thing, but it's not infeasible.

**Comment by JacquesCarette** [comment]
And my comment: GOOL should make it so 'by construction' instead of "making sure" post facto. Push the problem upstream!

**Comment by B-rando1** [comment]
I agree that it would be _easier_ to force the GOOL-writer to supply initial values. But at least in the naive way of doing this, I think this would have some nontrivial downsides, both for the GOOL-writer and the generated code. Consider this [Java code](https://github.com/JacquesCarette/Drasil/blob/9dd21479c945d212508d4945b3767946c751237b/code/stable/projectile/projectile_m_p_nol_b_u_v_d/src/java/Projectile/InputParameters.java#L15-L26) that we generate:

```java
public class InputParameters {
    public double v_launch;
    public double theta;
    public double p_target;
    
    /** \brief Initializes input object by reading inputs and checking physical constraints on the input
        \param filename name of the input file
    */
    public InputParameters(String filename) throws FileNotFoundException {
        this.get_input(filename);
        this.input_constraints();
    }
    
    /** \brief Reads input from a file with the given file name
        \param filename name of the input file
    */
    private void get_input(String filename) throws FileNotFoundException {
        Scanner infile;
        infile = new Scanner(new File(filename));
        infile.nextLine();
        this.v_launch = Double.parseDouble(infile.nextLine());
        infile.nextLine();
        this.theta = Double.parseDouble(infile.nextLine());
        infile.nextLine();
        this.p_target = Double.parseDouble(infile.nextLine());
        infile.close();
    }
```

If we force the GOOL-writer to use `stateVarDec` here instead of `stateVar`, then they have to supply meaningless default values, and GOOL has to generate code with those meaningless default values $-$ none of which affects the actual execution of the program. Worse, these default values don't actually make the code more "correct": if `get_input` fails to assign any of the class variables, then the program is still incorrect. Arguably it's worse, because the default value will prevent the Java compiler from noticing that the variable was never given a meaningful value.

I've been using the term "default value", which you said isn't what you're looking for; and maybe that's where part of my confusion comes from. But I don't see how GOOL's class constructors can be both correct by construction and flexible enough that the user isn't forced to provide meaningless defaults in cases like this. GOOL already has `Initializers`, where the user supplies a list of pairs of variables and values, and GOOL generates code at the top of the constructor assigning each variable its initial value. But clearly that wouldn't work for the above example. We can probably get a little more flexible than that, but getting to the point of calling other functions to perform the initialization while still being correct by construction seems like a big ask.

It's also a big ask to check post facto that each variable has been initialized, and not accessed before its initialization, by the end of the constructor. But I'd imagine there are established static analysis techniques for this sort of thing.

This comment is getting long, but to conclude $-$ if we can provide an interface for expressing, in whatever way best fits the use case, variable initialization in a way that is correct-by-construction, then that would be great. But if our interface isn't flexible enough, then the user will be forced to enter some meaningless initial values before writing custom code that overrides them.

**Comment by balacij** [comment]
On that code snippet, it looks like the code we should be generating something more like:

```java
public class InputParameters {
    public final double v_launch;
    public final double theta;
    public final double p_target;
    
    /** \brief Initializes input object by reading inputs and checking physical constraints on the input
        \param filename name of the input file
    */
    public InputParameters(double v_launch, double theta, double p_target) {
        this.v_launch = v_launch;
        this.theta = theta;
        this.p_target = p_target;

        this.input_constraints();
    }
    
    /** \brief Reads input from a file with the given file name
        \param filename name of the input file
    */
    public static InputParameters fromFile(String filename) throws FileNotFoundException {
        Scanner infile;
        infile = new Scanner(new File(filename));
        infile.nextLine();
        double v_launch = Double.parseDouble(infile.nextLine());
        infile.nextLine();
        double theta = Double.parseDouble(infile.nextLine());
        infile.nextLine();
        double p_target = Double.parseDouble(infile.nextLine());
        infile.close();

        return new InputParameters(v_launch, theta, p_target);
    }

    private void input_constraints() throws ValidationError { ... }
}
```


Or if we wanted to go all-in on Java conventions:

```java
public class InputParameters {
    private final double v_launch;
    private final double theta;
    private final double p_target;
    
    /** \brief Initializes input object by reading inputs and checking physical constraints on the input
        \param filename name of the input file
    */
    public InputParameters(double v_launch, double theta, double p_target) {
        this.v_launch = v_launch;
        this.theta = theta;
        this.p_target = p_target;

        this.input_constraints();
    }
    
    /** \brief Reads input from a file with the given file name
        \param filename name of the input file
    */
    public static InputParameters fromFile(String filename) throws FileNotFoundException {
        Scanner infile;
        infile = new Scanner(new File(filename));
        infile.nextLine();
        double v_launch = Double.parseDouble(infile.nextLine());
        infile.nextLine();
        double theta = Double.parseDouble(infile.nextLine());
        infile.nextLine();
        double p_target = Double.parseDouble(infile.nextLine());
        infile.close();

        return new InputParameters(v_launch, theta, p_target);
    }

    private void input_constraints() throws ValidationError { ... }

    public double getVLaunch() { return this.v_launch; }
    ...
}
```

**Comment by B-rando1** [comment]
We had a good discussion about this in our meeting.
- The code example I had in my [comment](https://github.com/JacquesCarette/Drasil/issues/2200#issuecomment-5035501736) is actually not very good code. Without default values, Swift would not accept it, because in its [Two-Phase Initialization model](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/initialization/#Two-Phase-Initialization), all instance variables must be defined before any instance method calls are made.
  - The first bit of code in Jason's [comment](https://github.com/JacquesCarette/Drasil/issues/2200#issuecomment-5037603759) is a better way of separating the concerns of file parsing and instance creation. Alternatively, we could inline the file parsing or use multiple constructors. Our Python renderer doesn't currently support multiple constructors though, so we'll have to find a solution to that if we want to go that route.
- Instead of a rigid set of default values for instance variables as in #5339, `stateVar` should require "evidence" that an instance variable will be initialized in a sane way. There are multiple forms of evidence we can talk about, e.g.:
  - A default value
  - A C++ style [Member initializer list](https://en.cppreference.com/cpp/language/constructor#Member_initializer_list)
  - An assignment to be put in the constructor
    - GOOL will need to topologically sort these
- Our notion of what information we need to give GOOL for a constructor will change through this. E.g. right now we have notions of `stateVar` vs `stateVarDef`; we have `Initializers` which work like in C++; and we have free-for-all "code" in our constructors. This will change, with "evidence" likely consolidating a lot of it into one place.

A side note, during the meeting @balacij asked again, "What is GOOL?" and @JacquesCarette said something to the effect of, "The design of GOOL is a linguistic act, not an abstraction exercise." That working definition is definitely helpful for framing our discussions.

This issue turned out to be much more of a rabbit-hole and I don't think it should be my main priority right now, but I'll try to get back to it once some of our refactors are done. We will also have to change some things in `drasil-code` first, so that it won't be trying to generate bad code that the new "evidence" system won't allow.
